README
This is a simplified graphic library I developed. 

This library can be used in lab sessions of languages C and Ada for CS-1 
and CS-2 students to make them more concrete. There are binary archives
for most of the popular UNIX-Like platforms.



ChangeLog
------------------------------------------------------------------------
r1208 | zalila | 2009-05-02 17:48:18 +0100 (Sam, 02 mai 2009) | 2 lines
Changed paths:
   M /work/libsimplified_graphics/mk.dev

	* Package changelog file

------------------------------------------------------------------------
r1207 | zalila | 2009-05-02 17:45:26 +0100 (Sam, 02 mai 2009) | 2 lines
Changed paths:
   A /work/libsimplified_graphics/ChangeLog

	* Changelog

------------------------------------------------------------------------
r1206 | zalila | 2009-05-02 12:01:47 +0100 (Sam, 02 mai 2009) | 2 lines
Changed paths:
   M /work/libsimplified_graphics/mk.dev

	* Add packaging option for sources

------------------------------------------------------------------------
r1184 | zalila | 2009-04-24 21:40:00 +0100 (Ven, 24 avr 2009) | 2 lines
Changed paths:
   M /work/libsimplified_graphics/graphiqu_enis_binding.ads

	* Align comment properly

------------------------------------------------------------------------
r1163 | zalila | 2009-03-05 20:43:25 +0100 (Jeu, 05 mar 2009) | 2 lines
Changed paths:
   M /work/libsimplified_graphics/graphiqu_enis.c
   M /work/libsimplified_graphics/mk.dev

	* Typos

------------------------------------------------------------------------
r1159 | zalila | 2009-03-02 23:15:15 +0100 (Lun, 02 mar 2009) | 2 lines
Changed paths:
   M /work/libsimplified_graphics/mk.dev

	* Portable test

------------------------------------------------------------------------
r1158 | zalila | 2009-03-02 23:12:19 +0100 (Lun, 02 mar 2009) | 2 lines
Changed paths:
   M /work/libsimplified_graphics/graphiqu_enis.c

	* Use a more portable way to puge the stdin buffer

------------------------------------------------------------------------
r1157 | zalila | 2009-03-02 22:56:10 +0100 (Lun, 02 mar 2009) | 2 lines
Changed paths:
   M /work/libsimplified_graphics/mk.dev

	* Better packaging

------------------------------------------------------------------------
r1154 | zalila | 2009-03-02 22:42:27 +0100 (Lun, 02 mar 2009) | 3 lines
Changed paths:
   M /work/libsimplified_graphics/Makefile
   M /work/libsimplified_graphics/graphiqu_enis.c
   M /work/libsimplified_graphics/mk.dev

	* Fixed a bug in Attente. Purge the stdin biuffer before waiting
	Better packaging (C and Ada are separated)

------------------------------------------------------------------------
r1152 | zalila | 2009-02-24 20:39:03 +0100 (Mar, 24 fév 2009) | 2 lines
Changed paths:
   M /work/libsimplified_graphics/graphiqu_enis.h
   M /work/libsimplified_graphics/graphiqu_enis_binding.ads
   M /work/libsimplified_graphics/mk.dev

	* minor reorganization

------------------------------------------------------------------------
r1150 | zalila | 2009-02-19 21:05:05 +0100 (Jeu, 19 fév 2009) | 2 lines
Changed paths:
   M /work/libsimplified_graphics/graphiqu_enis.h
   M /work/libsimplified_graphics/graphiqu_enis_binding.ads

	* Typos

------------------------------------------------------------------------
r1149 | zalila | 2009-02-19 20:56:19 +0100 (Jeu, 19 fév 2009) | 2 lines
Changed paths:
   M /work/libsimplified_graphics/mk.dev

	* Do not package the ada body

------------------------------------------------------------------------
r1148 | zalila | 2009-02-19 20:52:22 +0100 (Jeu, 19 fév 2009) | 2 lines
Changed paths:
   M /work/libsimplified_graphics/COPYING

	* Minor reformatting

------------------------------------------------------------------------
r1147 | zalila | 2009-02-19 20:50:08 +0100 (Jeu, 19 fév 2009) | 2 lines
Changed paths:
   A /work/libsimplified_graphics/COPYING
   M /work/libsimplified_graphics/graphiqu_enis.c
   M /work/libsimplified_graphics/graphiqu_enis.h
   M /work/libsimplified_graphics/graphiqu_enis_binding.adb
   M /work/libsimplified_graphics/graphiqu_enis_binding.ads
   M /work/libsimplified_graphics/mk.dev
   M /work/libsimplified_graphics/simple_test.c
   M /work/libsimplified_graphics/simple_test_ada.adb

	* Copyright

------------------------------------------------------------------------
r1146 | zalila | 2009-02-19 20:33:41 +0100 (Jeu, 19 fév 2009) | 2 lines
Changed paths:
   M /work/libsimplified_graphics/Makefile
   M /work/libsimplified_graphics/graphiqu_enis.h
   A /work/libsimplified_graphics/graphiqu_enis_binding.adb
   A /work/libsimplified_graphics/graphiqu_enis_binding.ads
   M /work/libsimplified_graphics/mk.dev
   A /work/libsimplified_graphics/simple_test_ada.adb

	* Binding Ada pour la bibliothèque graphique

------------------------------------------------------------------------
r1145 | zalila | 2009-02-15 12:03:18 +0100 (Dim, 15 fév 2009) | 5 lines
Changed paths:
   M /work/enst_c/monproj/geomet.c
   M /work/enst_c/monproj/geomet.h
   M /work/enst_c/monproj/main.c
   M /work/enst_c/monproj/outil_math.c
   M /work/enst_c/monproj/outil_math.h
   M /work/enst_c/monproj/utils.c
   M /work/enst_c/monproj/utils.h
   M /work/enst_c/monproj/var.h
   M /work/enst_c/tp_algo/exer6.c
   M /work/enst_c/tp_algo/ordonn2.c
   M /work/libsimplified_graphics/graphiqu_enis.c

	* libsimplified_graphics : Finiched the integration of
	  Dimensions_Fenetre
	* enst_c : Adapted the remaining graphical program to
	  Dimensions_Fenetre

------------------------------------------------------------------------
r1142 | zalila | 2009-02-10 12:20:22 +0100 (Mar, 10 fév 2009) | 3 lines
Changed paths:
   M /work/libsimplified_graphics/graphiqu_enis.c
   M /work/libsimplified_graphics/graphiqu_enis.h
   M /work/libsimplified_graphics/simple_test.c

	* A new function to make the windows size and the coordinate system
	separated and independant

------------------------------------------------------------------------
r1130 | zalila | 2009-02-03 20:20:09 +0100 (Mar, 03 fév 2009) | 4 lines
Changed paths:
   M /work/libsimplified_graphics/graphiqu_enis.c
   M /work/libsimplified_graphics/graphiqu_enis.h

	* Version 1.0
	* Activation du threads de raffraichissement sur Mac OS X en 
	utilisant les routines thread-safe de XLib


simple_test.c
/* $Id$ */

/*

  Copyright (C) 2009 Bechir Zalila <bechir.zalila@aist.enst.fr>
  Voir le fichier COPYING pour la licence

*/

#include <stdio.h>
#include <unistd.h>
#include "graphiqu_enis.h"

int main (void) {
  int i;
  float poly_1 [12] =
    {100, 0, 150, 150, 200, 150, 200, 200, 300, 200, 300, 0};
  float poly_2 [12] =
    {115, 10, 160, 140, 210, 140, 210, 190, 290, 190, 290, 10};
  float triangle [6];
  char size [5], msg [50];
  char tmpl [] = "Hello There @ '%s pixels'";

  /* Initialisation de l'environnement grahique avec les coordonnées
     extrêmales. */
  Initialisation_Graphique (-100, -150.7, 400.5, 300);

  /* Trace un trait entre les points (0, 0) et (50, 60) avec la
     couleur 6 (Brown) */
  Trait (0.0, 0.0, 50, 60, 6);

  /* Création d'une nouvelle couleur RGB (128, 128, 128) (Gris) et
     placement de cette couleur à la position 77. */
  CreateColorRGB (77, 128, 128, 128);

  /* Trace un trait entre les points (10, 0) et (60, 60) en utilisant
     la nouvelle couleur, 77. */
  Trait (10.0, 0.0, 60, 60, 77);

  /* Affichage d'une crois au point (0.5, 150) en utilisant la couleur
     noire (Black) */
  Croix (0.5, 150, Black);

  /* Affichage de 16 points avec les 16 premières couleurs */
  for (i = 0; i < 16; i++) {
    Point (100 + i * 10, 200.0, i);
  }

  /* Affichange de 16 cercles avec les 16 premieres couleurs et de 16
     autres disques avec un ordre de couleurs inverse. */
  for (i = 0; i < 16; i++) {
    Cercle (-80.0 + i * 30, 230.0, 15, i);
    Disque (-80.0 + i * 30, 260.0, 15, i);
  }

  /* Traçage d'un polygone en vert clair */
  
  DrawPoly (6, poly_1, LightGreen);

  /* Remplissage d'un second polygone légèrement plus petit que le
     premier, avec la couleur verte */
  FillPoly (6, poly_2, Green);

  /* Traçage de 16 triangles avec les 16 premières couleurs et
     remplissages de 16 autres triangles légèrement plus petis avec un
     ordre de couleurs inverse. */

  for (i = 0; i < 16; i++) {
    triangle [0] = -90.0 + 30 * i;
    triangle [1] = -90.0;
    triangle [2] = -60.0 + 30 * i;
    triangle [3] = -90.0;
    triangle [4] = -75.0 + 30 * i;
    triangle [5] = -60.5;
    DrawTriangle (triangle, i);
    triangle [0] = -85.0 + 30 * i;
    triangle [1] = -85.0;
    triangle [2] = -65.0 + 30 * i;
    triangle [3] = -85.0;
    triangle [4] = -75.0 + 30 * i;
    triangle [5] = -65.5;
    FillTriangle (triangle, 15 - i);
  }

  /* Traçage de 16 carrés avec les 16 premières couleurs et
     remplissages de 16 rectangles inscrits dans les carrés avec un
     ordre de couleurs inverse. */

  for (i = 0; i < 16; i++) {
    DrawRectangle (-90.0 + 30 * i, -120.0, -60.0 + 30 * i, -90.0, i);
    FillRectangle (-85.0 + 30 * i, -115.0, -65.0 + 30 * i, -90.0, 15 - i);
  }

  /* Affichage de quelques textes avec des tailles et des couleurs
     différentes. Traçage de rectangles ayant la même hauteur du texte
     à chaque fois. */
  
  Hauteur_Texte (8);
  itos (8, size);
  sprintf (msg, tmpl, size);
  EcritXY (50, 40, msg, 1);
  DrawRectangle (50, 40, 150, 48, 0);

  Hauteur_Texte (12);
  itos (12, size);
  sprintf (msg, tmpl, size);
  EcritXY (50, 50, msg, 5);
  DrawRectangle (50, 50, 150, 62, 0);
  Croix (50, 50, 1);

  Hauteur_Texte (10);
  itos (10, size);
  sprintf (msg, tmpl, size);
  EcritXY (50, 65, msg, 3);
  DrawRectangle (50, 65, 150, 75, 0);

  Hauteur_Texte (15);
  itos (15, size);
  sprintf (msg, tmpl, size);
  EcritXY (50, 0, msg, 4);
  DrawRectangle (50, 0, 150, 15, 0);
 
  /* Attente de la touche ENTREE de l'utilisateur */
  Attente ();

  /* Efface le contenu de la fenêtre */
  Efface ();

  /* Une seconde attente */
  Attente ();

  /* Destruction de la fenêtre graphique */
  Fin_Graphique ();

  return 0;
}


graphiqu_enis.c
/* $Id$ */

/*

  Copyright (C) 2009 Bechir Zalila <bechir.zalila@aist.enst.fr>
  Voir le fichier COPYING pour la licence

*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <X11/Xlib.h>
#include <assert.h>
#include <pthread.h>

#define MAX(x, y) (x > y ? x : y)
#define MIN(x, y) (x < y ? x : y)

#define CX(x) (w_width * ((x) - _xmin) / (_xmax - _xmin))
#define CY(y) (w_height - w_height * ((y) - _ymin) / (_ymax - _ymin))

char graphiqu_enis_version_string [] = "1.1";

int initialized = 0;

Display *d;  /* Le display */
int s;       /* L'écran */
Window w;    /* La fenêtre */
GC gc;       /* Le contexte graphique */
Pixmap bkp;  /* Une sauvegarde de ce qu'il y a sur la fenêtre */

unsigned int requested_width = 0, requested_height = 0;
/* Dimensions demandées par l'utilisateur. On peut par exemple avoir
   des coordonnées qui s'étendent par exemple entre -1 et 1 alors que
   la taille de la fenêtre est de 100 pixels */

unsigned int w_width, w_height;   /* Dimensions de la fenêtre */
float _xmin, _ymin, _xmax, _ymax; /* Sauvegardes des coordonnées */

float _cur_x, _cur_y; /* Position courante (utile pour l'écriture de
			 texte). */

XFontStruct* font_info = NULL; /* Police courante */

/* Nom générique de police redimensionnable. Le '%d' sera remplacé
   pour correspondre à la taille choisie par l'utilisateur. */
char* font_name_template 
  = "-adobe-helvetica-medium-r-normal--0-%d-75-75-p-0-iso8859-1";

int hauteur_texte = 10; /* Hauteur par défaut du texte */

void sync_font_info (void); 
/* Garantit la cohérence entre hauteur_texte et font_info */

Colormap colmap;  /* Table des couleurs système */

XColor TableCoul [256];
/* Table des couleurs. Les 16 premières valeurs seront remplies par
   les couleurs usuelles. Les 240 couleurs restantes forment un
   dégradé du vert au bleu-verdatre en passant par le jaune, le rouge,
   le violet et le bleu. Toutes les couleurs peuvent être modifiées
   par l'utilisateur. */

void init_couleurs (void);
/* Initialise les 256 couleurs */

pthread_t main_l;
/* Thread exécutant la boucle d'évènements */

void main_loop (void);
/* Boucle principale de la fenêtre graphique (sera exécutée dans un
   thread indépendant). */

void Graphiqu_ENIS_Version (void) {
  printf ("Bibliothèque graphique simplifiée de l'ENIS. Version %s\n",
	  graphiqu_enis_version_string);
  printf ("Copyright 2009 Bechir Zalila (bechir.zalila@aist.enst.fr)\n");
}

void Dimensions_Fenetre (int largeur, int hauteur) {
  requested_width = largeur;
  requested_height = hauteur;
}

void Initialisation_Graphique (float xmin, 
			       float ymin, 
			       float xmax, 
			       float ymax)
{
  XEvent e;
  unsigned int line_width = 1;
  int line_style = LineSolid;
  int cap_style = CapRound;
  int join_style = JoinRound;

  /* Vérifier que les dimensions données sont cohérentes */

  assert ((xmax - xmin) > 0);
  assert ((ymax - ymin) > 0);

  _xmin = xmin;
  _ymin = ymin;
  _xmax = xmax;
  _ymax = ymax;

  if (requested_width == 0) {
    w_width = (unsigned int) (xmax - xmin);
  } else {
    w_width = requested_width;
  }

  if (requested_height == 0) {
    w_height = (unsigned int) (ymax - ymin);
  } else {
    w_height = requested_height;
  }

  _cur_x = _xmin;
  _cur_y = _ymin;

  /* Message de version et de copyright */

  Graphiqu_ENIS_Version ();

  /* Initialisation de XLib avec le support multi-tâches */

  if (!XInitThreads ()) {
    printf ("Impossible d'initialiser XLib avec le support multi-tâches\n");
    exit (1);
  }

  /* Ouvrir la connexion avec le serveur graphique */
 
  d = XOpenDisplay (NULL);
  if (d == NULL) {
    printf ("Impossible d'ouvrir l'affichage\n");
    exit (1);
  }
  
  /* Verrouillage du display */

  XLockDisplay (d);
  
  /* Récupération de l'écran par défaut */
  
  s = XDefaultScreen (d);

  /* Création de la fenêtre */

  w = XCreateSimpleWindow (d, 
			   RootWindow (d, s), 
			   10, 
			   10, 
			   w_width + 2, 
			   w_height + 2,
			   1, 
			   BlackPixel (d, s), 
			   WhitePixel (d, s));  

  /* Choisir les événements qui nous intéressent */
  
  XSelectInput (d, w, StructureNotifyMask | ExposureMask | KeyPressMask);

  /* Création du contexte graphique */

  gc = XCreateGC (d, RootWindow (d, s), 0, NULL);
  XSetBackground (d, gc, WhitePixel (d, s));
  XSetForeground (d, gc, BlackPixel (d, s));

  /* Propriétés des lignes tracées */

  XSetLineAttributes (d, gc, 
		      line_width, line_style,
		      cap_style, join_style);

  /* Essayer de trouver une police de caractères */

  sync_font_info ();

  /* Création de la table de pixel qui sauvegarde la fenêtre et
     coloriage de son arrière plan en blanc. */
  
  bkp = XCreatePixmap (d,
		       RootWindow (d, s),
		       w_width + 2,
		       w_height + 2,
		       DefaultDepth (d, DefaultScreen (d)));
  XSetForeground (d, gc, WhitePixel (d, s));
  XFillRectangle (d, bkp, gc, 0, 0, w_width + 2, w_height + 2);
  XSetForeground (d, gc, BlackPixel (d, s));

  /* Initialisation des couleurs */

  init_couleurs ();

  /* Afficher la fenêtre */
  
  XMapWindow (d, w);
  XFlush (d);

  /* Se bloquer jusqu'à l'apparition de la fenêtre */

  while (1) {
    XNextEvent (d, &e);
    if (e.type == MapNotify) {
      break;
    }
  }

  /* Déverrouillage du disaply */

  XUnlockDisplay (d);

  initialized = 1;

  /* Lancement de l'event dispatch thread de la fenêtre. */

  pthread_create (&main_l, NULL, (void *) main_loop, NULL);
}

void Fin_Graphique (void)
{
  assert (initialized);

  XLockDisplay (d);
  XFreePixmap (d, bkp);
  XFreeGC (d, gc);
  XDestroyWindow (d, w);
  XFlush (d);
  XCloseDisplay (d);
  XUnlockDisplay (d);

  initialized = 0;
}

void Efface (void)
{
  assert (initialized);

  /* Traçage d'un rectangle blanc */

  XLockDisplay (d);
  XSetForeground (d, gc, WhitePixel (d, s));
  XFillRectangle (d, w, gc, 0, 0, w_width, w_height);
  XFillRectangle (d, bkp, gc, 0, 0, w_width, w_height);
  XFlush (d);
  XUnlockDisplay (d);
}

void Trait (float x1, 
	    float y1, 
	    float x2, 
	    float y2, 
	    int couleur)
{
  float _x1 = CX (x1);
  float _y1 = CY (y1);
  float _x2 = CX (x2);
  float _y2 = CY (y2);

  assert (initialized);

  XLockDisplay (d);
  XSetForeground (d, gc, TableCoul [couleur].pixel);
  XDrawLine (d, w, gc, _x1, _y1, _x2, _y2);
  XDrawLine (d, bkp, gc, _x1, _y1, _x2, _y2);
  XFlush (d);
  XUnlockDisplay (d);

}

void Croix (float x, 
	    float y, 
	    int couleur)
{
  float _x = CX (x);
  float _y = CY (y);
  int dim = 3;

  assert (initialized);

  XLockDisplay (d);
  XSetForeground (d, gc, TableCoul [couleur].pixel);
  XDrawLine (d, w, gc, _x - dim, _y - dim, _x + dim, _y + dim);
  XDrawLine (d, bkp, gc, _x - dim, _y - dim, _x + dim, _y + dim);
  XDrawLine (d, w, gc, _x - dim, _y + dim, _x + dim, _y - dim);
  XDrawLine (d, bkp, gc, _x - dim, _y + dim, _x + dim, _y - dim);
  XFlush (d);
  XUnlockDisplay (d);
}

void Point (float x, 
	    float y, 
	    int couleur)
{
  float _x = CX (x);
  float _y = CY (y);

  assert (initialized);

  XLockDisplay (d);
  XSetForeground (d, gc, TableCoul [couleur].pixel);
  XDrawPoint (d, w, gc, _x, _y);
  XDrawPoint (d, bkp, gc, _x, _y);
  XFlush (d);
  XUnlockDisplay (d);
}

void Cercle (float x, 
	     float y, 
	     int rayon, 
	     int couleur)
{
  float _x = CX (x);
  float _y = CY (y);

  assert (initialized);

  XLockDisplay (d);
  XSetForeground (d, gc, TableCoul [couleur].pixel);
  XDrawArc (d, w, gc,
	    _x - rayon, _y - rayon,  /* Point de départ */ 
	    2 * rayon, 2 * rayon,    /* Point d'arrivée */
	    0, 360*64);              /* Angles (en 64e de degré) */
  XDrawArc (d, bkp, gc,
	    _x - rayon, _y - rayon,  /* Point de départ */ 
	    2 * rayon, 2 * rayon,    /* Point d'arrivée */
	    0, 360*64);              /* Angles (en 64e de degré) */
  XFlush (d);
  XUnlockDisplay (d);
}

void Disque (float x, 
	     float y, 
	     int rayon, 
	     int couleur)
{
  float _x = CX (x);
  float _y = CY (y);

  assert (initialized);

  XLockDisplay (d);
  XSetForeground (d, gc, TableCoul [couleur].pixel);
  XFillArc (d, w, gc,
	    _x - rayon, _y - rayon,  /* Point de départ */ 
	    2 * rayon, 2 * rayon,    /* Point d'arrivée */
	    0, 360*64);              /* Angles (en 64e de degré) */
  XFillArc (d, bkp, gc,
	    _x - rayon, _y - rayon,  /* Point de départ */ 
	    2 * rayon, 2 * rayon,    /* Point d'arrivée */
	    0, 360*64);              /* Angles (en 64e de degré) */
  XFlush (d);
  XUnlockDisplay (d);
}

void DrawPoly (int nbr_points, 
	       float *poly_points, 
	       int couleur)
{
  int i;
  float _x1, _x2, _y1, _y2;
  
  assert (initialized);
  assert (nbr_points > 1);

  XLockDisplay (d);
  XSetForeground (d, gc, TableCoul [couleur].pixel);

  for (i = 0; i < nbr_points - 1; i++) {
    /* Tracer une ligne entre le point i et le point i + 1 */

    _x1 = CX (poly_points [2 * i]);
    _y1 = CY (poly_points [2 * i + 1]);
    _x2 = CX (poly_points [2 * i + 2]);
    _y2 = CY (poly_points [2 * i + 3]);
    XDrawLine (d, w, gc, _x1, _y1, _x2, _y2);
    XDrawLine (d, bkp, gc, _x1, _y1, _x2, _y2);
  }

  /* Relier le dernier point au premier point */
  
  _x1 = CX (poly_points [2 * (nbr_points - 1)]);
  _y1 = CY (poly_points [2 * nbr_points - 1]);
  _x2 = CX (poly_points [0]);
  _y2 = CY (poly_points [1]);
  XDrawLine (d, w, gc, _x1, _y1, _x2, _y2);
  XDrawLine (d, bkp, gc, _x1, _y1, _x2, _y2);
  
  XFlush (d);  
  XUnlockDisplay (d);
}

void FillPoly (int nbr_points, 
	       float *poly_points, 
	       int couleur)
{
  XPoint *pol;
  int i;

  assert (initialized);
  assert (nbr_points > 1);

  pol = (XPoint *) malloc (nbr_points * sizeof (XPoint));

  for (i = 0; i < nbr_points; i++) {
    pol [i].x = CX (poly_points [2 * i]);
    pol [i].y = CY (poly_points [2 * i + 1]);
  }

  XLockDisplay (d);
  XSetForeground (d, gc, TableCoul [couleur].pixel);
  XFillPolygon (d, w, gc, pol, nbr_points, Complex, CoordModeOrigin);
  XFillPolygon (d, bkp, gc, pol, nbr_points, Complex, CoordModeOrigin);
  XFlush (d);
  XUnlockDisplay (d);

  free (pol);
}

void DrawTriangle (float *poly_points, 
		   int couleur)
{
  DrawPoly (3, poly_points, couleur);
}

void FillTriangle (float *poly_points, 
		   int couleur)
{
  FillPoly (3, poly_points, couleur);
}

void DrawRectangle (float x1, 
		    float y1, 
		    float x2, 
		    float y2, 
		    int couleur)
{
  float _x1 = CX (x1),
    _x2 = CX (x2),
    _y1 = CY (y1),
    _y2 = CY (y2);
  
  float _x_M = MAX (_x1, _x2),
    _x_m = MIN (_x1, _x2),
    _y_M = MAX (_y1, _y2),
    _y_m = MIN (_y1, _y2);
    
  assert (initialized);

  XLockDisplay (d);
  XSetForeground (d, gc, TableCoul [couleur].pixel);
  XDrawRectangle (d, w, gc, _x_m, _y_m, (_x_M - _x_m), (_y_M - _y_m));
  XDrawRectangle (d, bkp, gc, _x_m, _y_m, (_x_M - _x_m), (_y_M - _y_m));  
  XFlush (d);
  XUnlockDisplay (d);
}

void FillRectangle (float x1, 
		    float y1, 
		    float x2, 
		    float y2, 
		    int couleur)
{
  float _x1 = CX (x1),
    _x2 = CX (x2),
    _y1 = CY (y1),
    _y2 = CY (y2);
  
  float _x_M = MAX (_x1, _x2),
    _x_m = MIN (_x1, _x2),
    _y_M = MAX (_y1, _y2),
    _y_m = MIN (_y1, _y2);
    
  assert (initialized);
  
  XLockDisplay (d);
  XSetForeground (d, gc, TableCoul [couleur].pixel);
  XFillRectangle (d, w, gc, _x_m, _y_m, (_x_M - _x_m), (_y_M - _y_m));
  XFillRectangle (d, bkp, gc, _x_m, _y_m, (_x_M - _x_m), (_y_M - _y_m));  
  XFlush (d);
  XUnlockDisplay (d);
}

void CreateColorRGB (int numero,
		     int R,
		     int G,
		     int B)
{
  assert ((numero >= 0) && (numero <= 255));
  
  TableCoul [numero].red   = R * 256;
  TableCoul [numero].green = G * 256;
  TableCoul [numero].blue  = B * 256;
  TableCoul [numero].flags = DoRed | DoGreen | DoBlue;
    
  XLockDisplay (d);
  XAllocColor (d, colmap, &TableCoul [numero]);
  XUnlockDisplay (d);
}

void Hauteur_Texte (float hauteur)
{
  assert (initialized);
  
  hauteur_texte = (int) hauteur;
  
  sync_font_info ();
}

void itos (int n, 
	   char s[])
{
  assert (initialized);
  sprintf (s, "%d", n);
}

void Deplace (float x1,
	      float y1)
{
  assert (initialized);
  
  _cur_x = x1;
  _cur_y = y1;
}

void EcritXY (float x, 
	      float y, 
	      char *chaine,
	      int couleur)
{
  int _x = CX (x),
    _y = CY (y);
  
  assert (initialized);

  XLockDisplay (d);
  XSetForeground (d, gc, TableCoul [couleur].pixel);
  XDrawString (d, w, gc, _x, _y, chaine, strlen (chaine));
  XDrawString (d, bkp, gc, _x, _y, chaine, strlen (chaine));
  XFlush (d);
  XUnlockDisplay (d);
}

void Ecrit (char *chaine,
	    int couleur)
{
  EcritXY (_cur_x, _cur_y, chaine, couleur);
}

void Attente (void)
{
  char c;

  assert (initialized);

  while ((c = fgetc (stdin)) != '\n' && c != EOF);

  c = '\0';
  while (c != '\n') {
    scanf ("%c", &c);
  }
}

void main_loop (void) {
  XEvent e;
  int evenement_present;

  while (initialized) {
    XLockDisplay (d);

    /* Voir s'il y a des évènements en attente */
    evenement_present = XPending (d);

    if (evenement_present) {
      /* S'il y a des evènements, on les récupère */

      XNextEvent (d, &e);
      XUnlockDisplay (d);
    } else {
      /* S'il n'y a pas de nouveaux évènement, on fait dormir le
	 thread pendant une brève durée (pour éviter d'occuper 100% du
	 CPU) et on continue la boucle. On estime que 0.1 seconde est
	 suffisante pour garantir un affichage agréable de la
	 fenêtre.
      */
      
      XUnlockDisplay (d);
      usleep (100000);
      continue;
    }

    if (e.type == Expose) {
      /* Si une partie de la fenêtre doit être redessinée, on la copie
	 depuis la table de pixel de sauvegarde. */

      XLockDisplay (d);
      XCopyArea (d, bkp, w, gc,
		 e.xexpose.x, e.xexpose.y,
		 e.xexpose.width, e.xexpose.height,
		 e.xexpose.x, e.xexpose.y);
      XUnlockDisplay (d);
    } 
  }
}

void init_couleurs (void) {
  char couleursPredef [16][15] = 
    {"black", 
     "blue", 
     "green", 
     "cyan", 
     "red",
     "darkmagenta", 
     "brown", 
     "lightgray", 
     "darkgray",
     "lightblue", 
     "lightgreen", 
     "lightcyan", 
     "orange", 
     "magenta", 
     "yellow", 
     "white"};
  
  int i;
  XColor exact;

  /* Pas de couleurs. On divise l'intervalle [0, 255] en 41
     sous-intervalles pour pouvoir avoir un dégradé réaliste sur 240
     couleurs. */
  int steps [42] = {0, 6, 12, 18, 25, 31, 37, 43, 50, 56, 62, 68, 75, 81, 
		    87, 93, 100, 106, 112, 118, 125, 131, 137, 143, 150, 156,
		    162, 168, 175, 181, 187, 193, 200, 206, 212, 218, 225, 
		    231, 237, 243, 249, 255};

  /* États de la machine à états qui génère le dégradé */
  enum p_states {i_green,         /* Augmente G à R et B nuls */
		 s_green_i_red,   /* Augmente R à G maximum et B nul */
		 s_red_d_green,   /* Diminue G à R maximum et B nul */
		 s_red_i_blue,    /* Augmente B à R maximum et G nul */
		 s_blue_d_red,    /* Diminue R à B maximum et G nul */
		 s_blue_i_green}; /* Augmente G à B maximum et R nul */

  enum p_states cur_state = i_green;
  int count, step;
  
  XLockDisplay (d);
  colmap = DefaultColormap (d, s);
  
  /* Construction des 16 premières couleurs */

  for (i = 0; i < 16; i++) {
    XAllocNamedColor (d, colmap, couleursPredef [i], &TableCoul [i], &exact);
  }
  XUnlockDisplay (d);

  /* Construction du dégradé sur les 240 positions restantes */

  count = 16;
  step = 12;

  while (count < 256) {
    switch (cur_state) 
      {
      case i_green :
	CreateColorRGB (count, steps [0], steps [step], steps [0]);
	if (step == 41) {
	  cur_state = s_green_i_red;
	  step = 0;
	} else {
	  step++;
	}
	break;

      case s_green_i_red :
	CreateColorRGB (count, steps [step], steps [41], steps [0]);
	if (step == 41) {
	  cur_state = s_red_d_green;
	  step = 41;
	} else {
	  step++;
	}
	break;

      case s_red_d_green :
	CreateColorRGB (count, steps [41], steps [step], steps [0]);
	if (step == 0) {
	  cur_state = s_red_i_blue;
	  step = 0;
	} else {
	  step--;
	}
	break;

      case s_red_i_blue :
	CreateColorRGB (count, steps [41], steps [0], steps [step]);
	if (step == 41) {
	  cur_state = s_blue_d_red;
	  step = 41;
	} else {
	  step++;
	}
	break;

      case s_blue_d_red :
	CreateColorRGB (count, steps [step], steps [0], steps [41]);
	if (step == 0) {
	  cur_state = s_blue_i_green;
	  step = 0;
	} else {
	  step--;
	}
	break;

      case s_blue_i_green :
	CreateColorRGB (count, steps [0], steps [step], steps [41]);
	if (step == 41) {
	  cur_state = i_green;
	  step = 0;
	  break;
	} else {
	  step++;
	}
	break;
      }
    count++;
  }
}

void sync_font_info (void) {
  char the_font_name [100];

  sprintf (the_font_name, font_name_template, hauteur_texte * 10);

  XLockDisplay (d);

  if (font_info != NULL) {
    XFreeFont (d, font_info);
  }

  font_info = XLoadQueryFont (d, the_font_name);

  if (font_info == NULL) {
    XUnlockDisplay (d);
    fprintf (stderr, "XLoadQueryFont: échec du chargement de"
	    " la police de caractère '%s'\n", the_font_name);
    exit (1);
  }
  else {
    XSetFont (d, gc, font_info->fid);
    XUnlockDisplay (d);
  }
}


demo.c
/* $Id$ */

/* Fichier de démonstration de la bibliothèque graphique simplifiée de
   l'ENIS. Fortement inspiré de la démonstration de la bibliothèque
   graphique de l'ENST:

   http://perso.telecom-paristech.fr/~danzart/fiches/graphique.html
*/

#include <stdio.h>
#include "graphiqu_enis.h"
#include <string.h>

int main(void) { 

  int n,i;
  char a[3];

  /* Definition de tableaux de reels qui contiennent les coordonnees
     (x,y) des points des polygones a tracer. */

  float polygone[18]={150,70,170,80,180,100,190,80,210,70,190,60,180,40,170,60};
  float triangle1[6]={150,150,220,250,170,240};
  float triangle2[6]={200,300,50,150,150,150};

  /* Ouverture d'une fenetre graphique definissant les coordonnees
     utilisateur. */

  Initialisation_Graphique(0,0,400,500);

  DrawRectangle(0,0,400,500,Red);
  DrawRectangle(10,10,390,435,Blue);
  
  /* Affichage d'un message*/

  Deplace(50,470);
  Hauteur_Texte (30.0);
  Ecrit("Dessins et couleurs", Orange);

  Hauteur_Texte(13.0);
  EcritXY(150,450,"Textes", LightBlue);

  /* Trace et Remplissage du premier polygone */

  DrawPoly(8,polygone,Orange);
  FillPoly(8,polygone,LightCyan);

  /* Trace et Remplissage d'un rectangle*/

  DrawRectangle(160,10,200,40,Blue);
  FillRectangle(160,10,200,40,LightBlue);

  /* Trace et Remplissage d'un cercle*/

  Cercle(180,300,10,Green);
  Disque(180,300,10,LightGreen);

  /* Affichage des 16 premieres couleurs*/

  Hauteur_Texte(10.0);
  for (n=0;n<16;n++){
    FillRectangle(360,50+20*n,375,50+20*n+15,n);
    DrawRectangle(360,50+20*n,375,50+20*n+15,Black);
    i=n;
    itos(i,a);
    EcritXY(340,57+20*n,a,Black);
  }

  /* Affichage des 256 couleurs disponibles */

  for (n=0;n<256;n++)
    Trait(50+n,400,50+n,430,n);

  for (n=0;n<256;n++){
    FillTriangle(triangle2,n);
  }

  /* Modification des composantes (Rouge,Vert,Bleu) de la couleur
     255. */

  CreateColorRGB(255,240,240,0);

  /* Trace et remplissage dans cette couleur du triangle tab2 */
  
  FillTriangle(triangle1,255);
  DrawTriangle(triangle1,LightBlue);
  
  Attente ();
  
  Fin_Graphique();

  printf("Fin du travail\n");
  return 0;
}
